### ISSUE TITLE
Review Terra data repository APIs

### ISSUE THREAD

**Issue by theathorn** [DataBiosphere/azul#1695]
From "[Using the Terra Data Repository](https://docs.google.com/document/d/1jCPVEapjJcJ3SMBd5ZNXcYkbD11nIoi5atsugcgCKWs/)":
  - https://jade.datarepo-dev.broadinstitute.org/swagger-ui.html

[Terra Data Repository Design](https://docs.google.com/document/d/1KGsrHt6lpiSL7e6LQ981BnmXAYVsddlZLqxIvDrT_Mk)

Google BigQuery APIs:
  - docs: https://cloud.google.com/bigquery/docs/
  - legacy SQL interface: https://cloud.google.com/bigquery/docs/reference/legacy-sql
  - standard SQL interface: https://cloud.google.com/bigquery/docs/reference/standard-sql/functions-and-operators

Check this box when you are done.

- [x] Noah
- [x] ~Daniel~
- [x] ~Hannes~

After checking the last box, move to Review pipeline.

**Comment by theathorn** [comment]
From Ruchi:
1. Most importantly -- I don't think you need the details on ingesting data, but more like searching for it, so exploring the APIs to query for data:

2. This is the example of direct BQ access you'll have to a dataset that is a good example: "datarepo_EncodeTest*"
https://console.cloud.google.com/bigquery?project=broad-jade-dev-data&p=broad-jade-dev-data&d=datarepo_EncodeTest3694de34_2908_4ff4_9327_1da0021828fb&page=dataset
(use the same burner account)

3. This is an example of the structure of a snapshot -- which is the data that is shared with a larger audience:
https://console.cloud.google.com/bigquery?project=broad-jade-dev-data&p=broad-jade-dev-data&d=encodefiletest_snapshot04f79019_f679_45ec_8319_0600edaf5319&page=dataset

And two assets inside of the snapshot, "file" and "donor"

3. A BQ schema for HCA is still in the works, but here's an example for Encode (I imagine certain portions of the schema will be reused-- like "donor" and "pipeline activity:
https://drive.google.com/open?id=1lsevWkjZQVqkVCAgloYcAgZjt3h9-bo8


Might be useful to get a discussion on the books, so once you've had a chance to explore, we can start figuring out:
1. What information would you like to see exposed in a "snapshot" of data?
2. Do we want one "asset" per type of entity in the databrowser -- project/samples/files?
3a) Will the browser support multiple "version" of a data model --not sure if my question is the right one
3b) How do you expect the assets to change as the data model evolves?

**Comment by noah-aviel-dove** [comment]
[GBQ JSON functions](https://cloud.google.com/bigquery/docs/reference/standard-sql/json_functions)

Example query that uses them in a `WHERE` clause:

```
SELECT content
FROM datarepo_broad_dsp_clinvar.datarepo_raw_clinical_assertion_bfca33d8_c0d3_48f7_908d_0f01048942d8
WHERE JSON_EXTRACT_SCALAR(content, "$.ClinVarAccession['@Type']") = "SCV"
LIMIT 20
```
`JSON_EXTRACT` returns a sub-object with keys. `JSON_EXTRACT_SCALAR` extracts primitive values. Keys that are not valid JS identifiers are escaped by using `['key']` syntax, such as `'@Type'` in the example above.

**Comment by hannes-ucsc** [comment]
@hannes-ucsc and @danielsotirhos will not do this on their own but instead get this knowleged from @noah-aviel-dove's demo tomorrow.

**Comment by noah-aviel-dove** [comment]
Followup questions from demo:
- [ ]  Contact Terra about apparent lack of endpoints for directly querying data
- [ ]  How to map snapshots to the correct GBQ tables?
- [ ]  How does pagination work in GBQ query results?

**Comment by noah-aviel-dove** [comment]
The name of a snapshot corresponds directly to the name of a GBQ dataset. Oddly, in cases where the name of the snapshot and the name of the dataset (as reported by the Terra API) do not match, it is the name of _snapshot_ that matches the name of a GBQ dataset. For example, the API reports that snapshot `93dc1e76-8f1c-4949-8f9b-07a087f3ce7b` has name `"WorkflowAnalysisSnapshot_bam"` and a source dataset with name `"1000GenomesDataset"`. However the GBQ dataset in `broad-jade-dev-data` is called `"WorkflowAnalysisSnapshot_bam"`, and there is no dataset called `"1000GenomesDataset"`. 

In short it appears we can retrieve the names of datasets and tables directly from the API for describing snapshots. The tables names are also provided by the API under `/api/repository/v1/snapshots{id}`.

**Comment by noah-aviel-dove** [comment]
Page size if specified using the optional `maxResults` query parameter. The response contains a `pageToken` field which can be provided as a query parameter to the subsequent request for accessing the next page of results. This info is from the synchronous [`query` endpoint](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/query).

**Comment by noah-aviel-dove** [comment]
Daniel Morgan: "You can use the TDR APIs to list the snapshots you have access to, and to get the metadata (note: not the same meaning as “metadata” in the HCA/DCP) about a particular snapshot such as its name and description. The only way to pull the data (including HCA/DCP metadata) out of a snapshot is to go through the BigQuery APIs."

Jeremy Hert: "Yep, seconding Dan's point here. One of Terra's design principles is "cloud-transparency" -- i.e. expose cloud APIs for consumers. Instead of putting another API in front of BQ or GCS, we broker access to those resources and get out of the way :)"

Therefore I wasn't missing anything just confused about the meaning of "metadata" in the Terra documentation. From Terra's perspective, HCA metadata is "data" and "metadata" is info on the snapshots/datasets that contain that HCA (meta)data.

**Comment by noah-aviel-dove** [comment]
I think everything's resolved now. Assigning to @hannes-ucsc for confirmation.

**Comment by melainalegaspi** [comment]
@hannes-ucsc : looks good
